### PR TITLE
[SPARK-43438][SQL] Error on missing input columns in `INSERT`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to raise an error when an user uses V1 `INSERT` without a list of columns, and the number of inserting columns doesn't match to the number of actual table columns.

At the moment Spark inserts data successfully in such case after the PR https://github.com/apache/spark/pull/41262 which changed the behaviour of Spark 3.4.x.

### Why are the changes needed?
1. To conform the SQL standard which requires the number of columns must be the same:
Screenshot 2023-08-07 at 11 01 27 AM

Apparently, the insertion below must not succeed:
```sql
spark-sql (default)> CREATE TABLE tabtest(c1 INT, c2 INT);
spark-sql (default)> INSERT INTO tabtest SELECT 1;
```

2. To have the same behaviour as **Spark 3.4**:
```sql
spark-sql (default)> INSERT INTO tabtest SELECT 1;
`spark_catalog`.`default`.`tabtest` requires that the data to be inserted have the same number of columns as the target table: target table has 2 column(s) but the inserted data has 1 column(s), including 0 partition column(s) having constant value(s).
```

### Does this PR introduce _any_ user-facing change?
Yes.

After the changes:
```sql
spark-sql (default)> INSERT INTO tabtest SELECT 1;
[INSERT_COLUMN_ARITY_MISMATCH.NOT_ENOUGH_DATA_COLUMNS] Cannot write to `spark_catalog`.`default`.`tabtest`, the reason is not enough data columns:
Table columns: `c1`, `c2`.
Data columns: `1`.
```

### How was this patch tested?
By running the modified tests:
```
$ build/sbt "test:testOnly *InsertSuite"
$ build/sbt "test:testOnly *ResolveDefaultColumnsSuite"
$ build/sbt -Phive "test:testOnly *HiveQuerySuite"
```

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 1]
```diff
@@ -88,37 +86,24 @@ object TableOutputResolver {
     if (actualExpectedCols.size < query.output.size) {
       throw QueryCompilationErrors.cannotWriteTooManyColumnsToTableError(
         tableName, actualExpectedCols.map(_.name), query)
+    } else if (actualExpectedCols.size > query.output.size && !byName) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:89]: shall we simply move this check to the else branch of `if (byName) ... else ...` below?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 2]
```diff
@@ -88,37 +86,24 @@ object TableOutputResolver {
...
-      // the column's default value. We need to pass `fillDefaultValue` as true here, if the
-      // `supportColDefaultValue` parameter is also true.
+      // the column's default value.
       reorderColumnsByName(
         tableName,
         query.output,
         actualExpectedCols,
         conf,
         errors += _,
-        fillDefaultValue = supportColDefaultValue)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:104]: why do we change this? It's by purpose that we don't enable default column value for v2 for now.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:104]: reverted back

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 3]
```diff
@@ -93,8 +93,7 @@ object TableOutputResolver {
     val errors = new mutable.ArrayBuffer[String]()
     val resolved: Seq[NamedExpression] = if (byName) {
       // If a top-level column does not have a corresponding value in the input query, fill with
-      // the column's default value. We need to pass `fillDefaultValue` as true here, if the
-      // `supportColDefaultValue` parameter is also true.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:97]: unnecessary change?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 4]
```diff
@@ -1027,13 +1031,13 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
       sql("create table t(i int, s bigint default 42, x bigint) using parquet")
-      sql("insert into t values(1)")
+      sql("insert into t(i) values(1)")
       checkAnswer(spark.table("t"), Row(1, 42L, null))
     }
     // The table has a partitioning column and a default value is injected.
     withTable("t") {
       sql("create table t(i boolean, s bigint, q int default 42) using parquet partitioned by (i)")
-      sql("insert into t partition(i='true') values(5, default)")
+      sql("insert into t partition(i='true') (s, q) values(5, default)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1040]: unnecessary change?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 5]
```diff
@@ -1495,14 +1499,14 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
-      sql("insert into t values(1)")
+      sql("insert into t(i) values(1)")
       checkAnswer(spark.table("t"), Row(1, 42, null))
     }
     // The table has a partitioning column and a default value is injected.
     withTable("t") {
       sql("create table t(i boolean, s bigint) using parquet partitioned by (i)")
       sql("alter table t add column q int default 42")
-      sql("insert into t partition(i='true') values(5, default)")
+      sql("insert into t partition(i='true') (s, q) values(5, default)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1509]: unnecessary change?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 6]
```diff
@@ -1517,7 +1521,7 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     withTable("t") {
       sql("create table t(i boolean default false) using parquet")
       sql("alter table t add column s bigint default 42")
-      sql("insert into t values(false, default), (default, 42)")
+      sql("insert into t(i, s) values(false, default), (default, 42)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1524]: ditto

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 7]
```diff
@@ -1255,7 +1255,7 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
     checkError(
       exception = intercept[AnalysisException] {
         sql(
-          """INSERT INTO TABLE dp_test PARTITION(dp)
+          """INSERT INTO TABLE dp_test PARTITION(dp) (key, value, dp)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:1258]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 8]
```diff
@@ -1027,13 +1031,13 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     }
     withTable("t") {
       sql("create table t(i int, s bigint default 42, x bigint) using parquet")
-      sql("insert into t values(1)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1030]: which Spark version added this test?

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1030]: Added by https://github.com/apache/spark/pull/41262 which haven't been released yet.

**Review by cloud-fan — APPROVED** [review]

**Review by dtenedor — APPROVED** [review]
LGTM, there's enough test coverage for INSERT commands with and without user-specified column lists, and with correct and incorrect numbers of attributes from the source table.

**Comment by MaxGekk** [comment]
Merging to master/3.5. Thank you, @cloud-fan and @dtenedor for review.
